buffer/src/main/java/io/netty/buffer/api/DefaultAllocator.java

```diff
@@ -0,0 +1,146 @@
+ * copy of the License at:
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package io.netty.buffer.api;
+
+import io.netty.util.internal.PlatformDependent;
+import io.netty.util.internal.SystemPropertyUtil;
+import io.netty.util.internal.logging.InternalLogger;
+import io.netty.util.internal.logging.InternalLoggerFactory;
+
+import java.util.Locale;
+import java.util.function.Supplier;
+
+import static io.netty.util.internal.ObjectUtil.checkNotNullWithIAE;
+import static io.netty.util.internal.PlatformDependent.directBufferPreferred;
+import static java.lang.Runtime.getRuntime;
+
+/**
+ * Accessor for the default, shared {@link BufferAllocator} instances.
+ * Two allocators are provided; one on-heap allocator, and one off-heap allocator.
+ * <p>
+ * These default allocators are configured via system properties.
+ * <p>
+ * These allocators cannot be {@linkplain Buffer#close() closed} directly.
+ * They will instead be disposed of when the {@link Runtime} is shutdown.
+ */
+public final class DefaultAllocator {
+    private static final InternalLogger logger = InternalLoggerFactory.getInstance(DefaultAllocator.class);
+    private static final BufferAllocator DEFAULT_PREFERRED_ALLOCATOR;
+    private static final BufferAllocator DEFAULT_ON_HEAP_ALLOCATOR;
+    private static final BufferAllocator DEFAULT_OFF_HEAP_ALLOCATOR;
+
+    static {
+        String allocType = SystemPropertyUtil.get(
+                "io.netty.allocator.type", PlatformDependent.isAndroid() ? "unpooled" : "pooled");
+        allocType = allocType.toLowerCase(Locale.US).trim();
+        boolean directBufferPreferred = directBufferPreferred();
+
+        BufferAllocator onHeap;
+        BufferAllocator offHeap;
+        if ("unpooled".equals(allocType)) {
+            onHeap = BufferAllocator.onHeapUnpooled();
+            offHeap = BufferAllocator.offHeapUnpooled();
+            logger.debug("-Dio.netty.allocator.type: {}", allocType);
+        } else if ("pooled".equals(allocType)) {
+            onHeap = BufferAllocator.onHeapPooled();
+            offHeap = BufferAllocator.offHeapPooled();
+            logger.debug("-Dio.netty.allocator.type: {}", allocType);
+        } else {
+            onHeap = BufferAllocator.onHeapPooled();
+            offHeap = BufferAllocator.offHeapPooled();
+            logger.debug("-Dio.netty.allocator.type: pooled (unknown: {})", allocType);
+        }
+        getRuntime().addShutdownHook(new Thread(() -> {
+            //noinspection EmptyTryBlock
+            try (onHeap; offHeap) {
+                // Left blank.
+            }
```

Comment:
Why not just call `close()` explicit ? I find this pattern used here confusing

Reply:
Try-with-resources automatically deals with exceptional cases correctly, like null references or close calls that throw.

Reply:
It becomes something like this, except I can't throw `Throwable`:

```
            Throwable throwable = null;
            try {
                onHeap.close();
            } catch (Throwable th) {
                throwable = th;
            }
            try {
                offHeap.close();
            } catch (Throwable th) {
                if (throwable == null) {
                    throwable = th;
                } else {
                    throwable.addSuppressed(th);
                }
            }
            if (throwable != null) {
                throw throwable;
            }
```